File: client/src/main/js/components/CourseSummary.js
import React from 'react';
import {
    Button,
    Card,
    CardBody,
    CardLink,
    CardTitle,
    Table
} from 'reactstrap';

export class CourseSummary extends React.Component {

    constructor(props) {
        super(props);
        this.saveResults = this.saveResults.bind(this);
    }

    saveResults() {
        //todo: Implement results aggregation and saving
    }

    render() {
        const tasksNames = this.props.course.tasks.sort();

        let studentIndex = 0;
        const studentsResults =
            this.props.tasks
                .flatMap(task => task.solutions)
                .groupBy(solution => solution.student)
                .map((solutions, student) =>
                    <tr>
                        <th scope="row">{studentIndex += 1}</th>
                        <td>{student}</td>
                        {
                            tasksNames
                                .map(task => {
                                    const solution = solutions.find(solution => solution.task === task);
                                    if (solution) {
                                        return (<td>{solution.score}</td>)
                                    } else {
                                        return (<td/>);
                                    }
                                })
                        }
                        <td>0</td>
                    </tr>
                )
                .valueSeq();

        return (
            <section className="course-summary">
                <Card>
                    <CardBody>
                        <CardTitle>Course summary</CardTitle>
                        <CardLink href={this.props.course.url}>Git repository</CardLink>
                        <Button className="save-results-btn"
                                onClick={this.saveResults}
                                color="primary"
                                outline
                        >
                            Save results
                        </Button>
                        <section className="course-stats">
                            <Table hover size="sm">
                                <thead>
                                <tr>
                                    <th>#</th>
                                    <th>Student</th>
                                    {tasksNames.map(name => <th>{name}</th>)}
                                    <th>Score</th>
                                </tr>
                                </thead>
                                <tbody>
                                {
                                    studentsResults.size > 0
                                        ? studentsResults
                                        :
                                        <td colSpan={tasksNames.size + 2}>There are no students on the course yet.</td>
                                }
                                </tbody>
                            </Table>
                        </section>
                    </CardBody>
                </Card>
            </section>
        )
    }
}

File: frontend/src/main/kotlin/org/flaxo/frontend/component/CourseSummary.kt
package org.flaxo.frontend.component

import kotlinx.html.TABLE
import kotlinx.html.TR
import kotlinx.html.ThScope
import kotlinx.html.js.onClickFunction
import org.flaxo.common.data.Course
import org.flaxo.common.data.CourseStatistics
import org.flaxo.common.data.Solution
import org.flaxo.frontend.github.githubProfileUrl
import react.RBuilder
import react.RComponent
import react.RProps
import react.dom.RDOMBuilder
import react.dom.a
import react.dom.button
import react.dom.div
import react.dom.h5
import react.dom.table
import react.dom.tbody
import react.dom.td
import react.dom.th
import react.dom.thead
import react.dom.tr

/**
 * Adds course summary.
 */
fun RBuilder.courseSummary(course: Course, courseStatistics: CourseStatistics) =
        child(CourseSummary::class) {
            attrs {
                this.course = course
                this.courseStatistics = courseStatistics
            }
        }

private class CourseSummaryProps(var course: Course, var courseStatistics: CourseStatistics) : RProps

private class CourseSummary(props: CourseSummaryProps) : RComponent<CourseSummaryProps, EmptyState>(props) {

    override fun RBuilder.render() {
        div(classes = "course-summary") {
            div(classes = "card") {
                div(classes = "card-body") {
                    h5(classes = "card-title") { +"Course summary" }
                    a(classes = "card-link", href = props.course.url) { +"Git repository" }
                    button(classes = "save-results-btn btn btn-outline-primary") {
                        attrs.onClickFunction = { saveScores() }
                        +"Save results"
                    }
                    courseStatisticsRefresh(props.course)
                    div(classes = "course-stats") {
                        table(classes = "table table-sm table-hover") {
                            tableHeader()
                            tableBody()
                        }
                    }

                }
            }
        }
    }

    private fun RDOMBuilder<TABLE>.tableHeader() {
        thead {
            tr {
                th(scope = ThScope.col) { +"#" }
                th(scope = ThScope.col) { +"Student" }
                props.courseStatistics.tasks
                        .map { it.branch }
                        .sorted()
                        .forEach { th(scope = ThScope.col) { +it } }
                th(scope = ThScope.col) { +"Score" }
            }
        }
    }

    private fun RDOMBuilder<TABLE>.tableBody() {
        tbody {
            props.courseStatistics.tasks
                    .flatMap { it.solutions }
                    .groupBy { it.student }
                    .toList()
                    .sortedBy { it.first }
                    .takeIf { it.isNotEmpty() }
                    ?.forEachIndexed { row, (student, solutions) ->
                        tr {
                            studentNumber(row)
                            studentName(student)
                            studentScores(solutions)
                            studentTotalScore(solutions)
                        }
                    }
                    ?: tr {
                        td {
                            attrs {
                                colSpan = (props.courseStatistics.tasks.size + 2).toString()
                            }
                            +"There are no students on the course yet."
                        }
                    }
        }
    }

    private fun RDOMBuilder<TR>.studentNumber(row: Int) =
            th(scope = ThScope.row) { +(row + 1).toString() }

    private fun RDOMBuilder<TR>.studentName(student: String) =
            td {
                a(classes = "github-profile", href = githubProfileUrl(student)) {
                    +student
                }
            }

    private fun RDOMBuilder<TR>.studentScores(solutions: List<Solution>) =
            props.courseStatistics.tasks
                    .map { it.branch }
                    .sorted()
                    .map { task -> solutions.find { it.task == task } }
                    .forEach { solution ->
                        solution?.also { td { +(it.score ?: 0).toString() } }
                                ?: td {}
                    }

    private fun RDOMBuilder<TR>.studentTotalScore(solutions: List<Solution>) =
            td {
                +solutions.map { it.score }
                        .map { it ?: 0 }
                        .map { it.toDouble() }
                        .let { scores -> (scores.sum() / (scores.size * 100)) * 100 }
                        .toInt()
                        .toString()
            }
}

private fun saveScores() {
    TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
}

